### PR TITLE
WIP: [feat] patch torch for improved performance in existing applications 

### PR DESCRIPTION
**Pull request by ClashLuke**
Implements #78\ 

- [ ] Run `pre-commit` before merging (WSL + Windows PyCharm is a bad combination here)
- [ ] Test if the code works
- [ ] If possible, add support for HuggingFace patching. (They reinvent the wheel a lot, so it's possible that they don't call torch.nn.Linear etc.).
- [ ] Benchmark performance of HuggingFace and PyTorch models before and after patching (maybe `benchmark_patched.py`?)

Feedback is appreciated, especially for the softmax wrapper.\
Should we add run-time assertions to ensure it's only called with [Batch, Sequence, Sequence] features?

### CODE REVIEW

File: xformers/triton/patch.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import typing
+
+import torch
+import torch.nn
+import torch.nn.functional
+
+from dropout import dropout
```
Comment (erip) [inline xformers/triton/patch.py:7]: Should these be absolute, e.g., `from xformers.triton.dropout import dropout`?

File: xformers/triton/patch.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+from fused_linear_layer import FusedLinear
+from layer_norm import FusedLayerNorm
+from softmax import softmax, log_softmax
+
+
+def _softmax(input: torch.Tensor, dim: int = -1, dtype=torch.float32):
+    if dim == -1:
+        return softmax(input)
```
Comment (dianaml0) [inline xformers/triton/patch.py:15]: Should this be limited to triton? For example there is sparse softmax as well available [here](https://github.com/facebookresearch/xformers/blob/861493c509de8bfc68895c9b5c58fbbc901d5176/xformers/components/attention/core.py#L96)

Comment (ClashLuke) [reply xformers/triton/patch.py:15]: I'm not sure I understand. Do you want to call the `_softmax` function instead? I wrote this wrapper to add transpose(), which we don't need with PyTorch softmax.

File: xformers/triton/patch.py (hunk 1 above)
Comment (ClashLuke) [reply xformers/triton/patch.py:7]: Good catch! I'll change that immediately. Thank you!

**Comment by erip** [comment]
It looks like you need to reformat `patch.py`. Running `black xformers/triton/patch.py` should autoformat.

**Comment by ClashLuke** [comment]
Unfortunately, I can't run `pre-commit` locally. If that's all that's left, could you run it on your machine?

File: xformers/triton/patch.py (hunk 2 above)
Comment (dianaml0) [reply xformers/triton/patch.py:15]: Seems like you're replacing `torch.nn.functional.softmax` with triton softmax only. Just suggesting you may want to also patch with sparse softmax if using sparse tensors or have a fallback when triton is not available, unless the focus of this is just triton? Then the way it is now seems fine.

**Comment by erip** [comment]
You don't need to add a precommit hook. Just install black, run the command, git add, commit, and push it.

File: xformers/triton/patch.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    return softmax(input.transpose(-1, dim)).transpose(-1, dim).to(dtype=dtype)
+
+
+def _log_softmax(input: torch.Tensor, dim: int = -1, dtype=torch.float32):
+    if dim == -1:
+        return log_softmax(input)
+    return log_softmax(input.transpose(-1, dim)).transpose(-1, dim).to(dtype=dtype)
+
+
+def _dropout(input: torch.Tensor, p: float = 0.5, training: bool = True, inplace: bool = False):
```
Comment (erip) [inline xformers/triton/patch.py:25]:
```suggestion
def _dropout(
    input: torch.Tensor, p: float = 0.5, training: bool = True, inplace: bool = False
):
```

File: xformers/triton/patch.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    if training:
+        return dropout(input, p)
+    return input * p
+
+
+def _layer_norm_wrapper(normalized_shape: typing.List[int], eps: float = 1e-5, elementwise_affine: bool = True,
+                        device=None, dtype=None):
+    return FusedLayerNorm(normalized_shape, elementwise_affine, eps).to(device=device, dtype=dtype)
```
Comment (erip) [inline xformers/triton/patch.py:33]:
```suggestion
def _layer_norm_wrapper(
    normalized_shape: typing.List[int],
    eps: float = 1e-5,
    elementwise_affine: bool = True,
    device=None,
    dtype=None,
):
    return FusedLayerNorm(normalized_shape, elementwise_affine, eps).to(
        device=device, dtype=dtype
    )
```

**Review by erip — CHANGES_REQUESTED** [review]
change formatting to satisfy black.

File: xformers/triton/patch.py (hunk 4 above)
Comment (ClashLuke) [reply xformers/triton/patch.py]: Thank you very much!

**Comment by erip** [comment]
The failure looks unrelated, but I'll take a look. If a maintainer can rerun the tests in the meantime, that would be lovely.

**Comment by blefaudeux** [comment]
> The failure looks unrelated, but I'll take a look. If a maintainer can rerun the tests in the meantime, that would be lovely.

sorry just seeing that now, I somehow missed it (ok, I was in a cabin without electricity at the time, could be related). Let me have a look

**Comment by blefaudeux** [comment]
Hey @ClashLuke thanks for the PR and sorry for having been so long to review, I somehow thought that it was a WIP and forgot to peek into it. I think it's a great idea but it's a bit hard to bullet proof, in that pytorch works absolutely all the time and it's a high bar to clear (the custom parts in xformers all assume the same conventions, which are maintained throughout the repo. Softmax is hardcoded to only ever apply to the last dimension for instance). Provided it's clear what the conventions are and if it's useful to people, I think that it's still great to have this. 

I would move that out of /triton and in the /utils if that makes sense, the /triton subfolder is about the triton implementations and it has some consequences for the code coverage for instance, would that be ok ? Thanks again !

p.s: pre-commit can be installed once ([see](https://pre-commit.com/)), you need to activate it once in the repo via `pre-commit install`, and after that all your commits will be auto linted and formatted, we need to update the contributing guide to show that
